hack/tools/janitor/vsphere.go

```diff
@@ -92,9 +105,106 @@ func newVSphereClients(ctx context.Context, input getVSphereClientInput) (*vSphe
 		return nil, err
 	}
 
+	fieldsManager, err := object.GetCustomFieldsManager(vimClient)
+	if err != nil {
+		return nil, err
+	}
+
+	viewManager := view.NewManager(vimClient)
+	finder := find.NewFinder(vimClient, false)
+
 	return &vSphereClients{
-		Vim:     vimClient,
-		Govmomi: govmomiClient,
-		Rest:    restClient,
+		Vim:           vimClient,
+		Govmomi:       govmomiClient,
+		Rest:          restClient,
+		FieldsManager: fieldsManager,
+		Finder:        finder,
+		ViewManager:   viewManager,
 	}, nil
 }
+
+const vSphereDeletionMarkerName = "capv-janitor-deletion-marker"
+
+func waitForTasksFinished(ctx context.Context, tasks []*object.Task, ignoreErrors bool) error {
+	for _, t := range tasks {
+		if err := t.Wait(ctx); !ignoreErrors && err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func getDeletionMarkerTimestamp(key int32, values []types.BaseCustomFieldValue) (*time.Time, error) {
+	// Find the value for the key
+	var b *types.BaseCustomFieldValue
+	for i := range values {
+		if values[i].GetCustomFieldValue().Key != key {
+			continue
+		}
+		b = &values[i]
+		break
+	}
+
+	// Key does not exist
+	if b == nil {
+		return nil, nil
+	}
+
+	value, ok := (*b).(*types.CustomFieldStringValue)
+	if !ok {
+		return nil, fmt.Errorf("cannot typecast %t to *types.CustomFieldStringValue", *b)
+	}
+
+	t, err := time.Parse(time.RFC3339, value.Value)
+	return &t, err
+}
+
+type managedElement struct {
+	entity  mo.ManagedEntity
+	element *list.Element
+}
+
+func recursiveList(ctx context.Context, inventoryPath string, govmomiClient *govmomi.Client, finder *find.Finder, viewManager *view.Manager, objectTypes ...string) ([]*managedElement, error) {
+	// Get the object at inventoryPath
+	objList, err := finder.ManagedObjectList(ctx, inventoryPath)
+	if err != nil {
+		return nil, err
+	}
+	if len(objList) != 1 {
+		return nil, errors.Errorf("expected to find object at managed object at path: %s", inventoryPath)
+	}
+
+	root := objList[0].Object.Reference()
```

Comment:
Is 0 the object at inventoryPath?

Reply:
Jep, we ensured that we found exactly 1 object, so the first item in objList is it.

Reply:
Ups :D

Reply:
Should have seen that